utils/misc.ml

```diff
@@ -694,3 +694,194 @@ module MakeHooks(M: sig
   let apply_hooks sourcefile intf =
     fold_hooks !hooks sourcefile intf
 end
+
+
+module Magic_number = struct
+  type version = int
+  type kind =
+      | Exec | Cmi | Cmo | Cma | Cmx | Cmxa | Cmxs | Cmt
+      | Ast_impl | Ast_intf
+
+  let all_kinds =
+    [ Exec; Cmi; Cmo; Cma; Cmx; Cmxa; Cmxs; Cmt; Ast_impl; Ast_intf ]
+
+  type raw = string
+  type info = {
+    kind: kind;
+    version: version;
+  }
+
+  let parse_kind : raw -> kind option = function
+    | "Caml1999X" -> Some Exec
+    | "Caml1999I" -> Some Cmi
+    | "Caml1999O" -> Some Cmo
+    | "Caml1999A" -> Some Cma
+    | "Caml1999Y" -> Some Cmx
+    | "Caml1999Z" -> Some Cmxa
+    | "Caml2007D" -> Some Cmxs
+    | "Caml2012T" -> Some Cmt
+    | "Caml1999M" -> Some Ast_impl
+    | "Caml1999N" -> Some Ast_intf
+    | _ -> None
+
+  let raw_kind : kind -> raw = function
+    | Exec -> "Caml1999X"
+    | Cmi -> "Caml1999I"
+    | Cmo -> "Caml1999O"
+    | Cma -> "Caml1999A"
+    | Cmx -> "Caml1999Y"
+    | Cmxa -> "Caml1999Z"
+    | Cmxs -> "Caml2007D"
+    | Cmt -> "Caml2012T"
+    | Ast_impl -> "Caml1999M"
+    | Ast_intf -> "Caml1999N"
+
+  let string_of_kind : kind -> string = function
+    | Exec -> "exec"
+    | Cmi -> "cmi"
+    | Cmo -> "cmo"
+    | Cma -> "cma"
+    | Cmx -> "cmx"
+    | Cmxa -> "cmxa"
+    | Cmxs -> "cmxs"
+    | Cmt -> "cmt"
+    | Ast_impl -> "ast_impl"
+    | Ast_intf -> "ast_intf"
+
+  let human_name_of_kind : kind -> string = function
+    | Exec -> "executable"
+    | Cmi -> "compile interface file"
+    | Cmo -> "bytecode object file"
+    | Cma -> "bytecode library"
+    | Cmx -> "native compilation unit description"
+    | Cmxa -> "static native library"
+    | Cmxs -> "dynamic native library"
+    | Cmt -> "cmt file"
+    | Ast_impl -> "serialized implementation AST"
+    | Ast_intf -> "serialized interface AST"
+
+  let kind_length = 9
+  let version_length = 3
+  let magic_length =
+    kind_length + version_length
+
+  type parse_error =
+    | Corrupted of string
+    | Not_a_magic_number of string
+
+  exception Parse_error of parse_error
+
+  let parse s : (info, parse_error) result =
+    if String.length s = magic_length then begin
+      let raw_kind = String.sub s 0 kind_length in
+      let raw_version = String.sub s kind_length version_length in
+      match parse_kind raw_kind with
+      | None -> Error (Not_a_magic_number s)
+      | Some kind ->
+          begin match int_of_string raw_version with
+          | exception _ -> Error (Corrupted s)
+          | version -> Ok { kind; version }
+          end
+    end
+    else begin
+      (* a header is "corrupted" if it starts like a valid magic number,
+         that is if its longest segment of length at most [kind_length]
+         is a prefix of [raw_kind kind] for some kind [kind] *)
```

Comment:
I kept the distinction, but I went back to `Truncated` as the constructor name. The `explain` function now special-cases `Truncated ""` and says that the file is empty (instead of just "truncated").